### PR TITLE
[REF] account: create 'account.reconciliation.widget' model

### PR DESCRIPTION
**Pull request by Gorash**
Contains widget reconciliation methods. This refactoring does not modify
the feature, some errors have been notified in the code but not modified.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Gorash** [comment]
I forgot to push enterprise, sorry

File: addons/account/models/widget_reconciliation.py [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+            if datum['type'] == 'account':
+                accounts = Account.browse(datum['id'])
+                accounts.mark_as_reconciled()
+
+    ####################################################
+    # Private
+    ####################################################
+
+    @api.model
+    # _get_move_lines_for_reconciliation
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/account/models/widget_reconciliation.py]: No. _ prefix is use for private methods and do not have access from rpc or other. It's private methods only called by this model

**Comment by Gorash** [comment]
https://github.com/odoo/enterprise/pull/1852

File: addons/account/models/widget_reconciliation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/account/models/widget_reconciliation.py]: oups sory, exact, the comment is wrong, I look the refactoring, sorry

File: addons/account/models/widget_reconciliation.py [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+        return Account_move_line
+
+    @api.model
+    # _get_reconciliation_proposition
+    def _get_move_line_reconciliation_proposition(self, account_id, partner_id=False):
+        """ Returns two lines whose amount are opposite """
+
+        Account_move_line = self.env['account.move.line']
+
+        # todo: remove crappy, self is an empty record set
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/account/models/widget_reconciliation.py]: old copy past, but I refactored this code and I forget to remove the comment

File: addons/account/models/widget_reconciliation.py [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+
+    @api.model
+    # model: 'account.move.line',
+    # method: 'get_move_lines_for_manual_reconciliation',
+    def get_move_lines_for_manual_reconciliation(self, account_id, partner_id=False, excluded_ids=None, str=False, offset=0, limit=None, target_currency_id=False):
+        """ Returns unreconciled move lines for an account or a partner+account, formatted for the manual reconciliation widget """
+
+        Account_move_line = self.env['account.move.line']
+        Account = self.env['account.account']
+        Currency = self.env['res.currency']
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/account/models/widget_reconciliation.py]: In many of the odoo methods, we declare the model variables at the top of the method in order to quickly know what are the links to the other models. I use the same guide lines.

File: addons/account/models/widget_reconciliation.py [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+    def _process_move_lines(self, move_line_ids, new_mv_line_dicts):
+        """ Create new move lines from new_mv_line_dicts (if not empty) then call reconcile_partial on self and new move lines
+
+            :param new_mv_line_dicts: list of dicts containing values suitable for account_move_line.create()
+        """
+        if len(move_line_ids) < 1 or len(move_line_ids) + len(new_mv_line_dicts) < 2:
+            raise UserError(_('A reconciliation must involve at least 2 move lines.'))
+
+        account_move_line = self.env['account.move.line'].browse(move_line_ids)
+        writeoff_lines = self.env['account.move.line']
```
[the remark this answers is not in the document]
Comment (Gorash) [reply addons/account/models/widget_reconciliation.py]: idem guide lines

**Comment by Gorash** [comment]
@oco-odoo ok ? (I'll merge the commit to adapt the pr in the previous commit after)

**Comment by oco-odoo** [comment]
@Gorash Yup! For me, it's alright :) (also the enteprise branch) Laura will test it, and then we can merge :)

File: addons/account/models/reconciliation_widget.py [hunk 5]
```diff
@@ -0,0 +1,782 @@
...
+                accounts = Account.browse(datum['id'])
+                accounts.mark_as_reconciled()
+
+    ####################################################
+    # Private
+    ####################################################
+
+    @api.model
+    # get_move_lines_for_reconciliation
+    def _get_move_lines(self, aml_accounts, partner_id, excluded_ids=None, search_str=False, offset=0, limit=None):
```
Comment (qdp-odoo) [inline addons/account/models/reconciliation_widget.py]: Can you motivate the removal of the parameters overlook_partner and additional_domain?

File: addons/l10n_generic_coa/data/account_bank_statement_demo.xml [hunk 6]
```diff
@@ -73,5 +73,18 @@
...
+            <field name="statement_id" ref="l10n_generic_coa.demo_bank_statement_1"/>
+            <field name="sequence">1</field>
+            <field name="name" eval="'SAJ/'+time.strftime('%Y')+'/002'"/>
+            <field name="journal_id" model="account.journal" search="[
+                ('type', '=', 'bank'),
+                ('company_id', '=', obj().env['res.company']._company_default_get('account.journal').id)]"/>
+            <field name="amount">750.0</field>
+            <field name="date" eval="time.strftime('%Y')+'-01-01'"/>
+        </record>
+
```
Comment (qdp-odoo) [inline addons/l10n_generic_coa/data/account_bank_statement_demo.xml]: do you use this demo data somewhere?

**Comment by mreficent** [comment]
Taking advantage of this refactoring, maybe you could also consider adding the companies like in https://github.com/odoo/odoo/pull/23070. It could be interesting and useful for people who work in more than one company. Thanks.

File: addons/account/models/reconciliation_widget.py (hunk 5 above)
Comment (Gorash) [reply addons/account/models/reconciliation_widget.py]: I made a modification, to have a method returning the domain. additional_domain and overlook_partner are never used in master or enterprise. If a community module needs to use it, just make an inheritance.

File: addons/l10n_generic_coa/data/account_bank_statement_demo.xml (hunk 6 above)
Comment (Gorash) [reply addons/l10n_generic_coa/data/account_bank_statement_demo.xml]: Yes, in the new tests.
I have add a tour to test different type of reconciliation. There will be more tests coming (according to information provided by Laura)

**Comment by Gorash** [comment]
@mreficent We will see this, but in a second time, so that people have already benefited from refactoring

File: addons/account/models/reconciliation_widget.py (hunk 5 above)
Comment (Gorash) [reply addons/account/models/reconciliation_widget.py]: overlook_partner is also use by _get_common_sql_query, I did not change this method, but it was never use in the get_move_lines_for_reconciliation method calls.

**Comment by qdp-odoo** [comment]
[IMP] account_batch_deposit: remove comments useful for forward-port

wrong module, Christophe ... can you rework the history when you're done with this branch? thanks

**Comment by Gorash** [comment]
@qdp-odoo of course, I copied the message from the enterprise branch without paying attention ^^

**Comment by qdp-odoo** [comment]
cool. let me know when you're done on this task so that I can close the PR

**Pull request merged**